pkg/manager/impl/artifact_manager.go

```diff
@@ -283,6 +308,97 @@ func (m *artifactManager) ListArtifacts(ctx context.Context, request *datacatalo
 	return &datacatalog.ListArtifactsResponse{Artifacts: artifactsList, NextToken: token}, nil
 }
 
+// UpdateArtifact updates the given artifact, currently only allowing the associated ArtifactData to be replaced. All
+// stored data will be overwritten in the underlying blob storage, no longer existing data (based on ArtifactData name)
+// will be deleted.
+func (m *artifactManager) UpdateArtifact(ctx context.Context, request *datacatalog.UpdateArtifactRequest) (*datacatalog.UpdateArtifactResponse, error) {
```

Comment:
Single comment on this function - IIUC we are updating by:
1. storing new ArtifactData (while updating the inmemory Artifact model)
2. deleting the existing existing data
3. updating the db Artifact
I think this could cause a corrupt system if there is a failure between 2 and 3. If we delete the existing ArtifactData and then fail to update the model then the next cache lookup will look for non-existent data. Does it make sense to switch steps 2 and 3? Then the biggest issue would be orphaned data in case of a failure - which could be easily cleaned up by searching the ArtifactData for an existing link.

Reply:
Basically, yeah.

We are:
1. Creating a "diff" between provided data and already existing one (based on artifact data name)
2. Creating "new" artifact data and overwriting existing one (in blob storage)
3. Deleting no longer existing artifact data (from blob storage)
4. Updating the database models (artifact and artifact data) to reflect the changes above

You're right about the broken state in case of an error, unfortunately we can't run in a transaction using the data store :sweat_smile:. Switching the order would work, I guess, however we might also end up with partially updated data (some of the new data was written successfully while parts of it haven't been updated yet) or artifact with both new _and_ old data (including results that should've been deleted already).

I guess that's still the best option for a partial failure like this (aside from completely nuking all cached data when encountering an error, which could potentially fail as well though).

Reply:
> unfortunately we can't run in a transaction using the data store
Could not agree more! I noticed the `CreateArtifact` implementation can have this same issue (orphaned data on failure). I don't think this is common issue now, but if we find it is we will have to implement some kind of garbage collection mechanism.

> I guess that's still the best option for a partial failure like this
Yeah, If you have other thoughts I'm very open. This is the best approach that I could think of - I think a partial update is preferential to a cache hit where the data does not actually exist.

Reply:
> Could not agree more! I noticed the `CreateArtifact` implementation can have this same issue (orphaned data on failure). I don't think this is common issue now, but if we find it is we will have to implement some kind of garbage collection mechanism.

Yeah, that could make sense - try to find orphaned data by comparing data store and database state. I'd like to keep that out of this PR/RFC implementation though just so we don't get too much feature creep.

> Yeah, If you have other thoughts I'm very open. This is the best approach that I could think of - I think a partial update is preferential to a cache hit where the data does not actually exist.

Definitely makes more sense than the current implementation. I'll see if I can come up with something else tomorrow, otherwise I'll change the order to your suggestion.

Reply:
@hamersaw Artifact data is now deleted after DB models have been created. I believe this should be enough for the main case we talked about (leaving "broken" artifact data entries without blob storage data).
Cleanup of potentially orphaned blob storage data is something for another PR, I'd say.

Reply:
I agree, thanks so much!